vm/ubpf_vm.c

```diff
@@ -1259,6 +1259,85 @@ ubpf_exec_ex(
             // Because we have already validated, we can assume that the type code is
             // valid.
             break;
+        case EBPF_OP_ATOMIC_STORE: {
+            BOUNDS_CHECK_STORE(8);
+            bool fetch = inst.imm & EBPF_ATOMIC_OP_FETCH;
+            volatile uint64_t* destination = (volatile uint64_t*)(reg[inst.dst] + inst.offset);
+            uint64_t value = reg[inst.src];
+            uint64_t result;
+            switch (inst.imm & EBPF_ALU_OP_MASK) {
+            case EBPF_ALU_OP_ADD:
+                result = UBPF_ATOMIC_ADD_FETCH(destination, value);
+                break;
+            case EBPF_ALU_OP_OR:
+                result = UBPF_ATOMIC_OR_FETCH(destination, value);
+                break;
+            case EBPF_ALU_OP_AND:
+                result = UBPF_ATOMIC_AND_FETCH(destination, value);
+                break;
+            case EBPF_ALU_OP_XOR:
+                result = UBPF_ATOMIC_XOR_FETCH(destination, value);
+                break;
+            case (EBPF_ATOMIC_OP_XCHG & ~EBPF_ATOMIC_OP_FETCH):
+                result = UBPF_ATOMIC_EXCHANGE(destination, value);
+                break;
+            case (EBPF_ATOMIC_OP_CMPXCHG & ~EBPF_ATOMIC_OP_FETCH):
+                result = UBPF_ATOMIC_COMPARE_EXCHANGE(destination, value, reg[0]);
+                break;
+            default:
+                vm->error_printf(stderr, "Error: unknown atomic opcode %d at PC %d\n", inst.imm, cur_pc);
+                return_value = -1;
+                goto cleanup;
+            }
+            if (fetch) {
+                if (inst.imm != EBPF_ATOMIC_OP_CMPXCHG) {
+                    reg[inst.src] = result;
```

Comment:
I spent some time determining what this code does. I am wondering if it might make sense to refactor slightly:
1. Make a variable that holds the `fetch_index`
2. Make it default to `inst.src`
3. Reset to `0` in the appropriate spot in the `case` statement above.
4. Use that index here instead of another conditional.

I am not sure that would make the code more or less readable, but I thought that I would offer it anyway.